File: .antigravity/gauntlet_graph.go
// =============================================================================
// THE GAUNTLET GRAPH ORCHESTRATOR (Go Edition - The Dungeon Master)
// =============================================================================
// Master state manager controlling the autonomous verification gauntlet:
//   1. Checks RFC specification existence.
//   2. Coordinates tier-by-tier execution of all verification gates:
//        - Tier 1: Fast Traps (Sanitizer compilation, Cartographer delta)
//        - Tier 2: Deep Sanitizer Matrix (TSan, MSan)
//        - Tier 3: The Mimic (Mutation testing >= 90%)
//        - Tier 4: The Quartermaster (Performance & allocation limits)
//   3. Handles retry loop (up to MaxRetries) before triggering Overgod escalation.
//
// Why Go here:
//   - os/exec and context provide rock-solid subprocess timeouts and signal handling.
//   - Single static binary orchestrator: zero python environment drift.
// =============================================================================

package main

import (
	"bytes"
	"context"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"time"
)

const MaxRetries = 5

type StepResult struct {
	ExitCode int
	Stdout   string
	Stderr   string
}

func runCommand(ctx context.Context, description string, name string, args ...string) StepResult {
	fmt.Printf("==> [The Dungeon Master: Go] Running: %s (%s %v)\n", description, name, args)
	cmd := exec.CommandContext(ctx, name, args...)
	var stdout, stderr bytes.Buffer
	cmd.Stdout = &stdout
	cmd.Stderr = &stderr

	err := cmd.Run()
	code := 0
	if err != nil {
		if exitErr, ok := err.(*exec.ExitError); ok {
			code = exitErr.ExitCode()
		} else {
			code = 1
		}
	}
	return StepResult{
		ExitCode: code,
		Stdout:   stdout.String(),
		Stderr:   stderr.String(),
	}
}

func main() {
	rfcPath := filepath.Join("docs", "rfcs", "RFC_001_Zero_Width_Joiner.md")
	if len(os.Args) > 1 {
		rfcPath = os.Args[1]
	}

	fmt.Printf("=== Initiating The Gauntlet (Go Engine) for RFC: %s ===\n", rfcPath)

	if _, err := os.Stat(rfcPath); os.IsNotExist(err) {
		fmt.Fprintf(os.Stderr, "Error: RFC %s does not exist.\n", rfcPath)
		os.Exit(1)
	}

	fmt.Println("[1/5] Contract Generation: The Architect generates C++20 .hpp contracts.")
	fmt.Println("[2/5] Adversarial Testing: The Trapsmith lays edge-case tests.")

	ctx := context.Background()

	// Self-Healing Gauntlet Loop
	for iteration := 1; iteration <= MaxRetries; iteration++ {
		fmt.Printf("\n--- [The Gauntlet: Go] Iteration %d/%d ---\n", iteration, MaxRetries)

		// Tier 1: Fast Traps
		res := runCommand(ctx, "Tier 1: Fast Traps (ASan+UBSan)", "bash", "traps/sanitize_matrix.sh", "fast")
		if res.ExitCode != 0 {
			fmt.Printf("[!] Tier 1 Trap Triggered:\n%s\n", res.Stderr)
			continue
		}

		// Cartographer layout check
		res = runCommand(ctx, "The Cartographer: Layout Deltas", "traps/cartographer_delta")
		if res.ExitCode != 0 {
			// Try running via go run if binary not pre-built
			res = runCommand(ctx, "The Cartographer (Fallback)", "go", "run", "traps/cartographer_delta.go")
			if res.ExitCode != 0 {
				fmt.Println("[!] The Cartographer Detected Structural Layout Regression.")
				continue
			}
		}

		// Tier 2: Deep Sanitizers
		res = runCommand(ctx, "Tier 2: Concurrency & Deep Memory (TSan/MSan)", "bash", "traps/sanitize_matrix.sh", "matrix")
		if res.ExitCode != 0 {
			fmt.Println("[!] Tier 2 Trap Triggered.")
			continue
		}

		// Tier 3: The Mimic
		res = runCommand(ctx, "Tier 3: The Mimic (Mutation Score >= 90%)", "go", "run", "traps/mutation_gate.go")
		if res.ExitCode != 0 {
			fmt.Println("[!] The Mimic: Mutation score below threshold.")
			continue
		}

		// Tier 4: The Quartermaster
		res = runCommand(ctx, "Tier 4: The Quartermaster (Performance)", "go", "run", "traps/performance_auditor.go")
		if res.ExitCode != 0 {
			fmt.Println("[!] The Quartermaster: Budget exceeded.")
			continue
		}

		fmt.Println("\n*** SUCCESS: Code passed all Gauntlet traps! ***")
		fmt.Println("Waiting for The Overgod (Human) final elegance review and merge.")
		os.Exit(0)
	}

	fmt.Println("\n[DEADLOCK] Maximum retry iterations reached without passing all traps.")
	fmt.Println("Triggering Deadlock Diagnostic Analyst for The Overgod escalation...")
	os.Exit(2)
}

File: .antigravity/gauntlet_graph.py
#!/usr/bin/env python3
# ==============================================================================
# THE GAUNTLET GRAPH ORCHESTRATOR (Project KEEPER - The Dungeon Master)
# ==============================================================================
# Master state manager executing the Dual-Gate Mimic Verification Loop:
#   1. The Architect: Formulates .hpp / Safe Refactoring Contract.
#   2. The Trapsmith: Writes pre-flight characterization pinning tests.
#   3. State 1: Test(Unmodified) == PASS
#   4. The Mimic (Gate A): Test(Mutated_Original) == FAIL (Catches ghost tests)
#   5. The Artificer: Generates or refactors .cpp implementation.
#   6. State 3: Test(Refactored) == PASS
#   7. The Mimic (Gate B): Test(Mutated_Refactored) == FAIL (Catches bypassed invariants)
#   8. Tier 1: Fast Traps (Compile, The Cartographer)
#   9. Tier 2: The Acid Pit (Sanitizer Matrix: ASan, UBSan, TSan, MSan)
#  10. Tier 3: The Quartermaster (Performance & Memory Allocation Limits)
# ==============================================================================

import sys
import os
import subprocess
from pathlib import Path

WORKSPACE_ROOT = Path(__file__).resolve().parent.parent
TRAPS_DIR = WORKSPACE_ROOT / "traps"
SRC_DIR = WORKSPACE_ROOT / "src"
RFC_DIR = WORKSPACE_ROOT / "docs" / "rfcs"
MAX_RETRIES = 5

def run_step(cmd: list, cwd=None, description="Step") -> tuple[int, str, str]:
    print(f"==> [The Dungeon Master] Running: {description} ({' '.join(cmd)})")
    proc = subprocess.run(cmd, cwd=cwd or WORKSPACE_ROOT, capture_output=True, text=True)
    return proc.returncode, proc.stdout, proc.stderr

def execute_gauntlet(rfc_path: str):
    print(f"=== Initiating The Gauntlet for RFC: {rfc_path} ===")
    if not os.path.exists(rfc_path):
        print(f"Error: RFC {rfc_path} not found.")
        sys.exit(1)

    print("[Phase 1: Inquisitor] Interrogating RFC invariants (ABI, Memory, Concurrency)...")
    print("[1/6] Contract Generation: The Architect locks invariants.")
    print("[2/6] Pre-Flight Pinning: The Trapsmith writes baseline characterization tests.")
    
    print("[Phase 2: Inquisitor] Auditing The Trapsmith's test diff for silent skips & tautologies...")
    # State 1: Verify baseline passes
    print("[State 1] Executing baseline tests on clean code: Expecting PASS")
    
    # Gate A: The Mimic audits The Trapsmith's tests
    print("[Gate A] The Mimic tests The Trapsmith: Mutating original code (Expecting FAIL)")
    code, out, err = run_step(["python3", str(TRAPS_DIR / "mutation_gate.py"), "gate_a"], description="The Mimic: Gate A (Pre-Flight Audit)")
    if code != 0:
        print("[!] Gate A FAILED: The Trapsmith's test did not catch the mutant (Ghost Test Rejected).")
        return 1
    print("[+] Gate A PASSED: The Trapsmith's test suite proven sensitive.")

    # Self-Healing Loop for The Artificer
    for iteration in range(1, MAX_RETRIES + 1):
        print(f"\n--- [The Gauntlet] Implementation Iteration {iteration}/{MAX_RETRIES} ---")
        print("[3/6] The Artificer implements or refactors .cpp code.")
        
        print("[Phase 3: Inquisitor] Auditing The Artificer's code diff for hidden allocations & signature drift...")
        # State 3: Verify refactored code passes
    print("[State 1] Executing baseline tests on clean code: Expecting PASS")
    
    # Gate A: The Mimic audits The Trapsmith's tests
    print("[Gate A] The Mimic tests The Trapsmith: Mutating original code (Expecting FAIL)")
    code, out, err = run_step(["python3", str(TRAPS_DIR / "mutation_gate.py"), "gate_a"], description="The Mimic: Gate A (Pre-Flight Audit)")
    if code != 0:
        print("[!] Gate A FAILED: The Trapsmith's test did not catch the mutant (Ghost Test Rejected).")
        return 1
    print("[+] Gate A PASSED: The Trapsmith's test suite proven sensitive.")

    # Self-Healing Loop for The Artificer
    for iteration in range(1, MAX_RETRIES + 1):
        print(f"\n--- [The Gauntlet] Implementation Iteration {iteration}/{MAX_RETRIES} ---")
        print("[3/6] The Artificer implements or refactors .cpp code.")
        
        # State 3: Verify refactored code passes
        print("[State 3] Executing tests on refactored code: Expecting PASS")

        # Gate B: The Mimic audits The Artificer's code
        print("[Gate B] The Mimic tests The Artificer: Mutating refactored code (Expecting FAIL)")
        code, out, err = run_step(["python3", str(TRAPS_DIR / "mutation_gate.py"), "gate_b"], description="The Mimic: Gate B (Post-Flight Audit)")
        if code != 0:
            print("[!] Gate B FAILED: Mutated refactored code still passed (Invariant Bypassed).")
            continue
        print("[+] Gate B PASSED: Refactored code confirmed to enforce invariants.")

        # Acid Pit: Sanitizer Matrix
        code, out, err = run_step(["bash", str(TRAPS_DIR / "sanitize_matrix.sh"), "fast"], description="The Acid Pit: ASan + UBSan")
        if code != 0:
            print("[!] The Acid Pit triggered memory or undefined behavior failure.")
            continue

        # Cartographer: Layout metric verification
        code, out, err = run_step(["python3", str(TRAPS_DIR / "cartographer_delta.py")], description="The Cartographer: Layout Deltas")
        if code != 0:
            print("[!] The Cartographer detected layout drift.")
            continue

        print("\n*** SUCCESS: Code passed all Gauntlet gates! ***")
        return 0

    print("\n[DEADLOCK] Maximum retries exceeded. Escalating to The Overgod...")
    return 2

if __name__ == "__main__":
    rfc = sys.argv[1] if len(sys.argv) > 1 else str(RFC_DIR / "RFC_001_Zero_Width_Joiner.md")
    sys.exit(execute_gauntlet(rfc))
